File: pruning/algo.py
def solveNQueens(self, n):
    if n < 1:
        return []
    self.result = []
    self.cols = set(); self.pie = set(); self.na = set()
    self.DFS(n, 0, [])
    return self.gen(n)

def DFS(self, n, row, cur):
    if row >= n:
        self.result.append(cur)
        return
    for col in range(n):
        if col in self.cols or row + col in self.pie or row - col in self.na:
            continue
        self.cols.add(col)
        self.pie.add(row + col)
        self.na.add(row - col)
        self.DFS(n, row + 1, cur + [col])
        self.cols.remove(col)
        self.pie.remove(row + col)
        self.na.remove(row - col)

def gen(self, n):
    board = []
    for res in self.result:
        for i in res:
            board.append("." * i + "Q" + "." * (n - i - 1))
    return [board[i:i+n] for i in range(0, len(board), n)]

def totalNQueens(self, n):
    if n < 1: return []
    self.total = 0
    self.cols = set(); self.pie = set(); self.na = set()
    self.DFS(n, 0, [])
    return self.total

def DFS(self, n, row, cur):
    if row >= n:
        self.total = self.total + 1
        return
    for col in range(n):
        if col in self.cols or row + col in self.pie or row - col in self.na:
            continue
        self.cols.add(col)
        self.pie.add(row + col)
        self.na.add(row - col)
        self.DFS(n, row + 1, cur + [col])
        self.cols.remove(col)
        self.pie.remove(row + col)
        self.na.remove(row - col)

File: pruning/algo.java
public void solveSudoku(char[][] board) {
    if (board == null || board.length == 0) {
        return;
    }
    solve(board);
}

public boolean solve(char[][] board) {
    for (int i = 0; i < board.length; i++) {
        for (int j = 0; j < board[0].length; j++) {
            if (board[i][j] == '.') {
                for (char c = '1'; c <= '9'; c++) {
                    if (isValid(board, i, j, c)) {
                        board[i][j] = c;
                        if (solve(board)) {
                            return true;
                        } else {
                            board[i][j] = '.';
                        }
                    }
                }
                return false;
            }
        }
        return true;
    }
}

public boolean isValid(char[][] board, int row, int col, char c) {
    for (int i = 0; i < 9; i++) {
        if (board[i][col] != '.' && board[i][col] == c) return false;
        if (board[row][i] != '.' && board[row][i] == c) return false;
        if (board[3*(row/3)+i/3][3*(col/3)+i%3] != '.' && board[3*(row/3)+i/3][3*(col/3)+i%3] == c) return false;
    }
    return true;
}

public boolean isValidSudoku(char[][] board) {
    for (int i = 0; i < board.length; i++) {
        for (int j = 0; j < board[0].length; j++) {
            if (board[i][j] != '.' && !isValid(board, i, j)) {
                return false;
            }
        }
    }
    return true;
}

public boolean isValid(char[][] board, int row, int col) {
    char c = board[row][col];
    for (int i = 0; i < 9; i++) {
        if (i != row && board[i][col] != '.' && board[i][col] == c) return false;
        if (i != col && board[row][i] != '.' && board[row][i] == c) return false;
        if (3*(row/3)+i/3 != row && 3*(col/3)+i%3 != col && board[3*(row/3)+i/3][3*(col/3)+i%3] != '.' && board[3*(row/3)+i/3][3*(col/3)+i%3] == c) return false;
    }
    return true;
}